windows/src/developer/kmcmpdll/versioning.cpp

```diff
@@ -1,45 +1,36 @@
 
-#include <string.h>
 #include <ctype.h>
+#include <stdio.h>
+
+#include <string.h>
 #include <windows.h>
 
 // Keyman includes
 
-#include <keyman32.h>
-
+#include <keyman64.h>
 #include <compfile.h>
 #include <compiler.h>
 #include <comperr.h>
-#include <vkeys.h>
-
 
-//TVersion FVersionInfo;
+BOOL AddCompileString(LPSTR buf);
+BOOL AddCompileMessage(DWORD msg);
 
-/* How to determine the minimum file version:
+BOOL CheckKeyboardFinalVersion(PFILE_KEYBOARD fk) {
+  char buf[128];
 
-  1. Minimum file version is always 0x0500
-  2. When first "call" statement is referenced, min file version moves to 0x0501
-  3. If keyboard contains VERSION 5.1, then min file version moves to 0x0510
+  if (fk->dwFlags & KF_AUTOMATICVERSION) {
+    wsprintf(buf, "The compiler has assigned a minimum engine version of %x.%x based on features used in this keyboard", (fk->version & 0xFF00) >> 8, fk->version & 0xFF);
+    AddCompileString(buf);
+  }
 
-  Ensuring that features are not used if not available in VERSION x.xx statement:
-
-  -- all older features available in VERSION 5.0 or earlier
-  -- call available in VERSION 5.01 or later
-  -- context(n), LHS context, LHS index, vkeys and deadkeys in stores,
-     named code constants, mnemonic vs positional settings, ethnologue codes,
-	 "old character position" system store, all only available in VERSION 5.1
-
-
-  -- 0x0500:		Keyman 5.0.100.0
-  -- 0x0501:		Keyman 5.0.112.0
-  -- 0x0510:		Keyman 5.1.1xx.0
-*/
-
-void SetMinFileVersion(int version)
-{
-	FVersionInfo.MinVersion = version;
+  return TRUE;
 }
 
-void UpdateKeyboardVersion(PWSTR p)
-{
+BOOL VerifyKeyboardVersionF(PFILE_KEYBOARD fk, DWORD ver) {
+  if (fk->dwFlags & KF_AUTOMATICVERSION) {
+    fk->version = max(fk->version, ver);
+    return TRUE;
```

Comment:
Making sure I get this - this function is called at multiple points during compilation, allowing this function to incrementally compare its current minimum version against that needed to support each feature being checked if `KF_AUTOMATICVERSION` is specified.  If insufficient, the version is increased, resulting in the minimum necessary version being held in `fk->version` post-compilation.

Reply:
Correct